File: designPatterns/Structural_patterns/Flyweight/python/FlyweightFactory.py
from Flyweight import Flyweight


class FlyweightFactory:
    def __init__(self, initial_flyweights):
        self.flyweights = {}
        for state in initial_flyweights:
            self.flyweights[self.get_key(state)] = Flyweight(state)

    def get_key(self, state):
        return "_".join(state)

    def get_flyweight(self, shared_state):
        key = self.get_key(shared_state)

        if not key in self.flyweights:
            print("FlyweightFactory: Can't find a flyweight, creating new one.")
            self.flyweights[key] = Flyweight(shared_state)
        else:
            print("FlyweightFactory: Reusing existing flyweight.")

        return self.flyweights[key]

    def list_flyweights(self):
        count = len(list(self.flyweights.keys()))
        print(f"\nFlyweightFactory: I have {count} flyweights:")
        for key in self.flyweights:
            print(key)

File: designPatterns/Structural_patterns/Flyweight/javascript/FlyweightFactory.js
const Flyweight = require("./Flyweight");

/**
 * The Flyweight Factory creates and manages the Flyweight objects. It ensures
 * that flyweights are shared correctly. When the client requests a flyweight,
 * the factory either returns an existing instance or creates a new one, if it
 * doesn't exist yet.
 */
class FlyweightFactory {
  constructor(initialFlyweights) {
    this.flyweights = {};
    for (const state of initialFlyweights) {
      this.flyweights[this.getKey(state)] = new Flyweight(state);
    }
  }

  // Returns a Flyweight's string hash for a given state.
  getKey(state) {
    return state.join("_");
  }

  // Returns an existing Flyweight with a given state or creates a new one.
  getFlyweight(sharedState) {
    const key = this.getKey(sharedState);

    if (!(key in this.flyweights)) {
      console.log("FlyweightFactory: Can't find a flyweight, creating new one.");
      this.flyweights[key] = new Flyweight(sharedState);
    } else {
      console.log("FlyweightFactory: Reusing existing flyweight.");
    }

    return this.flyweights[key];
  }

  listFlyweights() {
    const count = Object.keys(this.flyweights).length;
    console.log(`\nFlyweightFactory: I have ${count} flyweights:`);
    for (const key in this.flyweights) {
      console.log(key);
    }
  }
}

module.exports = FlyweightFactory;
